File: cloudfunctions/plan/checks.js
'use strict'

function buildChecks({ plan, places, itinerary, weatherByPlace }) {
  const issues = []
  const add = (level, code, title, detail = '') => {
    issues.push({ level, code, title, detail })
  }

  if (!plan.date) add('warn', 'plan.missing_date', '未填写日期')

  if (!places.length) {
    add('warn', 'places.empty', '还没有添加地点')
    return issues
  }

  const missingAdcode = places.filter((p) => !(String(p.adcode || '').trim()))
  if (missingAdcode.length) {
    add('warn', 'places.missing_adcode', '部分地点缺少 adcode', '缺少 adcode 的地点无法稳定查询天气。')
  }

  const assigned = new Set(itinerary.map((it) => it.place_id).filter(Boolean))
  const unassigned = places.filter((p) => !assigned.has(p.id))
  if (unassigned.length) {
    add('warn', 'itinerary.unassigned', '有地点未安排时间段', `未安排数量：${unassigned.length}`)
  }

  const bySlot = { morning: [], afternoon: [], evening: [] }
  for (const it of itinerary) {
    if (bySlot[it.time_slot] && it.place_id) bySlot[it.time_slot].push(it.place_id)
  }
  for (const [slot, ids] of Object.entries(bySlot)) {
    if (ids.length >= 4) {
      add('warn', 'itinerary.too_many', '单个时间段地点过多', `${slot} 安排了 ${ids.length} 个地点，可能过于紧凑。`)
    }
  }

  const budget = plan.budget
  const people = plan.people_count || 1
  if (typeof budget === 'number' && budget > 0) {
    const perPerson = budget / Math.max(1, Number(people) || 1)
    if (perPerson < 80) {
      add(
        'warn',
        'plan.low_budget',
        '人均预算偏低',
        `人均约 ¥${Math.round(perPerson)}，可能需要减少跨区移动或选择更省钱的活动。`
      )
    }
  }

  const rainPlaces = []
  for (const p of places) {
    const w = weatherByPlace[p.id]
    if (w && typeof w === 'object') {
      const s = String(w.status || '')
      if (s.includes('雨')) rainPlaces.push(p.name || '地点')
    }
  }
  if (rainPlaces.length) {
    add(
      'warn',
      'weather.rain_risk',
      '雨天出行风险',
      '部分地点天气为“雨”，建议准备雨具并考虑室内备选：' + rainPlaces.slice(0, 5).join('、')
    )
  }

  if (!issues.length) {
    add('info', 'ok', '未发现明显问题', '可以继续完善地点备注与时间安排。')
  }
  return issues
}

module.exports = { buildChecks }

File: backend/checks.py
from __future__ import annotations

from typing import Any


def build_checks(
    *,
    plan: dict[str, Any],
    places: list[dict[str, Any]],
    itinerary: list[dict[str, Any]],
    weather_by_place: dict[str, Any],
) -> list[dict[str, Any]]:
    issues: list[dict[str, Any]] = []

    def add(level: str, code: str, title: str, detail: str = ""):
        issues.append(
            {
                "level": level,  # info|warn
                "code": code,
                "title": title,
                "detail": detail,
            }
        )

    # plan basics
    if not plan.get("date"):
        add("warn", "plan.missing_date", "未填写日期")

    # places
    if not places:
        add("warn", "places.empty", "还没有添加地点")
        return issues

    missing_adcode = [p for p in places if not (p.get("adcode") or "").strip()]
    if missing_adcode:
        add(
            "warn",
            "places.missing_adcode",
            "部分地点缺少 adcode",
            "缺少 adcode 的地点无法稳定查询天气。",
        )

    # itinerary completeness
    assigned = {it.get("place_id") for it in itinerary if it.get("place_id")}
    unassigned = [p for p in places if p.get("id") not in assigned]
    if unassigned:
        add(
            "warn",
            "itinerary.unassigned",
            "有地点未安排时间段",
            f"未安排数量：{len(unassigned)}",
        )

    # too many in a slot
    by_slot: dict[str, list[str]] = {"morning": [], "afternoon": [], "evening": []}
    for it in itinerary:
        slot = it.get("time_slot")
        pid = it.get("place_id")
        if slot in by_slot and pid:
            by_slot[slot].append(pid)
    for slot, ids in by_slot.items():
        if len(ids) >= 4:
            add(
                "warn",
                "itinerary.too_many",
                "单个时间段地点过多",
                f"{slot} 安排了 {len(ids)} 个地点，可能过于紧凑。",
            )

    # budget heuristic
    budget = plan.get("budget")
    people = plan.get("people_count") or 1
    if isinstance(budget, int) and budget > 0:
        per_person = budget / max(1, int(people))
        if per_person < 80:
            add(
                "warn",
                "plan.low_budget",
                "人均预算偏低",
                f"人均约 ¥{per_person:.0f}，可能需要减少跨区移动或选择更省钱的活动。",
            )

    # rain risk heuristic (AMap base weather text)
    rain_places = []
    for p in places:
        w = weather_by_place.get(p.get("id"))
        if isinstance(w, dict):
            status = str(w.get("status") or "")
            if "雨" in status:
                rain_places.append(p.get("name") or "地点")
    if rain_places:
        add(
            "warn",
            "weather.rain_risk",
            "雨天出行风险",
            "部分地点天气为“雨”，建议准备雨具并考虑室内备选：" + "、".join(rain_places[:5]),
        )

    if not issues:
        add("info", "ok", "未发现明显问题", "可以继续完善地点备注与时间安排。")

    return issues

